// picgo/dist/src/lib/PluginHandler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cross_spawn_1 = __importDefault(require("cross-spawn"));
class PluginHandler {
    constructor(ctx) {
        this.ctx = ctx;
    }
    async install(plugins, proxy = '', env) {
        plugins = plugins.map((item) => 'picgo-plugin-' + item);
        const result = await this.execCommand('install', plugins, this.ctx.baseDir, proxy, env);
        if (!result.code) {
            plugins.forEach((plugin) => {
                this.ctx.pluginLoader.registerPlugin(plugin);
            });
            this.ctx.log.success('插件安装成功');
            this.ctx.emit('installSuccess', {
                title: '插件安装成功',
                body: plugins
            });
        }
        else {
            const err = `插件安装失败，失败码为${result.code}，错误日志为${result.data}`;
            this.ctx.log.error(err);
            this.ctx.emit('installFailed', {
                title: '插件安装失败',
                body: err
            });
        }
    }
    async uninstall(plugins) {
        plugins = plugins.map((item) => 'picgo-plugin-' + item);
        const result = await this.execCommand('uninstall', plugins, this.ctx.baseDir);
        if (!result.code) {
            plugins.forEach((plugin) => {
                this.ctx.pluginLoader.unregisterPlugin(plugin);
            });
            this.ctx.log.success('插件卸载成功');
            this.ctx.emit('uninstallSuccess', {
                title: '插件卸载成功',
                body: plugins
            });
        }
        else {
            const err = `插件卸载失败，失败码为${result.code}，错误日志为${result.data}`;
            this.ctx.log.error(err);
            this.ctx.emit('uninstallFailed', {
                title: '插件卸载失败',
                body: err
            });
        }
    }
    async update(plugins, proxy = '', env) {
        plugins = plugins.map((item) => 'picgo-plugin-' + item);
        const result = await this.execCommand('update', plugins, this.ctx.baseDir, proxy, env);
        if (!result.code) {
            this.ctx.log.success('插件更新成功');
            this.ctx.emit('updateSuccess', {
                title: '插件更新成功',
                body: plugins
            });
        }
        else {
            const err = `插件更新失败，失败码为${result.code}，错误日志为${result.data}`;
            this.ctx.log.error(err);
            this.ctx.emit('updateFailed', {
                title: '插件更新失败',
                body: err
            });
        }
    }
    execCommand(cmd, modules, where, proxy = '', env = {}) {
        const registry = this.ctx.getConfig('registry');
        return new Promise((resolve, reject) => {
            let args = [cmd].concat(modules).concat('--color=always').concat('--save');
            if (registry) {
                args = args.concat(`--registry=${registry}`);
            }
            if (proxy) {
                args = args.concat(`--proxy=${proxy}`);
            }
            try {
                const npm = cross_spawn_1.default('npm', args, { cwd: where, env: Object.assign({}, process.env, env) });
                let output = '';
                npm.stdout.on('data', (data) => {
                    output += data;
                }).pipe(process.stdout);
                npm.stderr.on('data', (data) => {
                    output += data;
                }).pipe(process.stderr);
                npm.on('close', (code) => {
                    if (!code) {
                        resolve({ code: 0, data: output });
                    }
                    else {
                        reject({ code: code, data: output });
                    }
                });
                // for users who haven't installed node.js
                npm.on('error', (err) => {
                    this.ctx.log.error(err);
                    this.ctx.log.error('NPM is not installed');
                    this.ctx.emit('failed', 'NPM is not installed');
                });
            }
            catch (e) {
                this.ctx.log.error(e);
                this.ctx.emit('failed');
            }
        });
    }
}
exports.default = PluginHandler;
